src/containers/ShareTopicModal.js

```diff
@@ -39,47 +51,187 @@ export default class ShareTopicModal extends React.Component {
 
   render () {
     const { onCancel, tagName, community, dispatch, pending,
-      tagInvitationEditor: { recipients, error, success } } = this.props
+      tagInvitationEditor: { recipients, recipient, error, success } } = this.props
+
+    const { copied } = this.state
+
     const joinUrl = get(community, 'beta_access_code')
       ? communityTagJoinUrl(community, tagName)
       : null
 
     const setError = text => dispatch(updateTagInvitationEditor('error', text))
 
-    const update = (field) => event =>
-      dispatch(updateTagInvitationEditor(field, event.target.value))
+    const update = (field) => value =>
+      dispatch(updateTagInvitationEditor(field, value))
+
+    const addRecipient = recipient => {
+      update('recipients')(recipients.concat(recipient))
+    }
+
+    const updateRecipient = text =>
+      update('recipient')(text)
+
+    const removeRecipient = recipient => {
+      const test = recipient.id
+        ? r => r.id !== recipient.id
+        : r => r !== recipient
+      update('recipients')(filter(test, recipients))
+    }
 
     const submit = () => {
       setError(null)
 
-      let emails = parseEmailString(recipients)
-      if (isEmpty(emails)) return setError('Enter at least one email address.')
+      if (isEmpty(recipients)) return setError('Enter at least one email address or user.')
+
+      const users = flow(
+        filter(r => r.id),
+        map('id'))(recipients)
+
+      const emails = filter(r => !r.id, recipients)
 
       let badEmails = emails.filter(email => !validator.isEmail(email))
       if (some(badEmails)) return setError(`These emails are invalid: ${badEmails.join(', ')}`)
 
-      dispatch(sendCommunityTagInvitation(community.id, tagName, {emails}))
+      dispatch(sendCommunityTagInvitation(community.id, tagName, {
+        users,
+        emails: emails.join(',')
+      }))
     }
 
-    return <Modal title='Invite new members' subtitle={`Invite people to join ${community.name} and follow #${tagName}`} id='share-topic' onCancel={onCancel}>
+    const copyLink = joinUrl
+      ? () => {
```

Comment:
can you make this a one-liner, i.e. `() => copy(joinUrl) && this.setState({copied: true})`? looks weird to have the whole thing embedded in a ternary expression otherwise